src/transport/TCP.h

```diff
@@ -0,0 +1,235 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP Connection object that maintains TCP connections.
+ *      It binds to any avaiable local addr and port and begins listening.
+ */
+
+#ifndef __TCPTRANSPORT_H__
+#define __TCPTRANSPORT_H__
+
+#include <algorithm>
+#include <utility>
+
+#include <core/CHIPCore.h>
+#include <inet/IPAddress.h>
+#include <inet/IPEndPointBasis.h>
+#include <inet/InetInterface.h>
+#include <transport/Base.h>
+
+namespace chip {
+namespace Transport {
+
+/** Defines listening parameters for setting up a TCP transport */
+class TcpListenParameters
+{
+public:
+    explicit TcpListenParameters(Inet::InetLayer * layer) : mLayer(layer) {}
+    TcpListenParameters(const TcpListenParameters &) = default;
+    TcpListenParameters(TcpListenParameters &&)      = default;
+
+    Inet::InetLayer * GetInetLayer() { return mLayer; }
+
+    Inet::IPAddressType GetAddressType() const { return mAddressType; }
+    TcpListenParameters & SetAddressType(Inet::IPAddressType type)
+    {
+        mAddressType = type;
+
+        return *this;
+    }
+
+    uint16_t GetListenPort() const { return mListenPort; }
+    TcpListenParameters & SetListenPort(uint16_t port)
+    {
+        mListenPort = port;
+
+        return *this;
+    }
+
+    InterfaceId GetInterfaceId() const { return mInterfaceId; }
+    TcpListenParameters & SetInterfaceId(InterfaceId id)
+    {
+        mInterfaceId = id;
+
+        return *this;
+    }
+
+private:
+    Inet::InetLayer * mLayer         = nullptr;               ///< Associated inet layer
+    Inet::IPAddressType mAddressType = kIPAddressType_IPv6;   ///< type of listening socket
+    uint16_t mListenPort             = CHIP_PORT;             ///< TCP listen port
+    InterfaceId mInterfaceId         = INET_NULL_INTERFACEID; ///< Interface to listen on
+};
+
+/**
+ * Packets scheduled for sending once a connection has been established.
+ */
+struct PendingPacket
+{
+    PeerAddress peerAddress;             // where the packet is being sent to
+    System::PacketBuffer * packetBuffer; // what data needs to be sent
+};
+
+/** Implements a transport using TCP. */
+class DLL_EXPORT TCPBase : public Base
+{
+    /**
+     *  The State of the TCP connection
+     */
+    enum class State
+    {
+        kNotReady    = 0, /**< State before initialization. */
+        kInitialized = 1, /**< State after class is listening and ready. */
+    };
+
+public:
+    TCPBase(Inet::TCPEndPoint ** activeConnectionsBuffer, size_t bufferSize, PendingPacket * packetBuffers,
+            size_t packetsBuffersSize) :
+        mActiveConnections(activeConnectionsBuffer),
+        mActiveConnectionsSize(bufferSize), mPendingPackets(packetBuffers), mPendingPacketsSize(packetsBuffersSize)
+    {
+        PendingPacket emptyPending{
+            .peerAddress  = PeerAddress::Uninitialized(),
```

Comment:
Designated initializers in C++ are non-standard until C++20, no?  I'm a little surprised this compiled on all our platforms, actually.

Reply:
If it compiles everywhere, should I bother to update it as this is quite readable?